proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java

```diff
@@ -173,12 +179,72 @@ public boolean handle(AvailableCommands commands) {
       }
     }
 
-    server.getEventManager().fire(
-        new PlayerAvailableCommandsEvent(serverConn.getPlayer(), rootNode))
-        .thenAcceptAsync(event -> playerConnection.write(commands), playerConnection.eventLoop());
+    server
+        .getEventManager()
+        .fire(new PlayerAvailableCommandsEvent(serverConn.getPlayer(), rootNode))
+        .thenAcceptAsync(
+            event -> {
+              commands.setRootNode(
```

Comment:
To avoid unnecessarily processing the backend server commands (which don't have a requires  predicate), I would move this filtering to https://github.com/MrIvanPlays/Velocity/blob/dfce2ea2d39d9eb6350db8705c07c705b22d4a52/proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java#L175 (just above this, where the proxy commands are injected). This also avoids the cloning if the `announceProxyCommands` config is set to `false`

Reply:
If I move it there, the nodes that the event adds wouldn't be captured and the method wouldn't work as intended. 
Additionally, we can remove this event and enforce `BrigadierCommand` if someone wants to modify the root node.

Reply:
Imo, nodes added via `PlayerAvailableCommandsEvent` should always be sent, even if the source doesn't pass the predicate.

Reply:
> Imo, nodes added via `PlayerAvailableCommandsEvent` should always be sent, even if the source doesn't pass the predicate.

how come

Reply:
If a plugin forcefully adds a command node to the tree, it should be sent to the client regardless of the `requires` return. The plugin can register a command to use this mechanic. The event is mostly intended to remove or edit nodes.

Reply:
> If a plugin forcefully adds a command node to the tree, it should be sent to the client regardless of the `requires` return. The plugin can register a command to use this mechanic. The event is mostly intended to remove or edit nodes.

Sure, I will move the filtering then